MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTest.c

```diff
@@ -0,0 +1,1208 @@
+
+  //
+  // Build BlockIo media structure
+  //
+  Device->Media.MediaId          = 0;
+  Device->Media.RemovableMedia   = FALSE;
+  Device->Media.MediaPresent     = TRUE;
+  Device->Media.LogicalPartition = FALSE;
+  Device->Media.ReadOnly         = FALSE;
+  Device->Media.WriteCaching     = FALSE;
+  // Device->Media.IoAlign          = Private->PassThruMode.IoAlign;
+
+  // Lbads                   = NamespaceData->LbaFormat[LbaFmtIdx].Lbads;
+  // Device->Media.BlockSize = (UINT32)1 << Lbads;
+  Device->Media.BlockSize = (UINT32)(1 << 9); // 512 byte sector size
+
+  Device->Media.LastBlock                     = 0x4000; // NamespaceData=>Nsze
+  Device->Media.LogicalBlocksPerPhysicalBlock = 1;
+  Device->Media.LowestAlignedLba              = 1;
+
+  Device->BlockIo.Revision    = EFI_BLOCK_IO_PROTOCOL_REVISION2;
+  Device->BlockIo.Media       = &Device->Media;
+  Device->BlockIo.ReadBlocks  = NvmeBlockIoReadBlocks;
+  Device->BlockIo.WriteBlocks = NvmeBlockIoWriteBlocks;
+
+  Device->BlockIo2.Media         = &Device->Media;
+  Device->BlockIo2.ReadBlocksEx  = NvmeBlockIoReadBlocksEx;
+  Device->BlockIo2.WriteBlocksEx = NvmeBlockIoWriteBlocksEx;
+
+  Device->MediaSanitize.Revision    = MEDIA_SANITIZE_PROTOCOL_REVISION;
+  Device->MediaSanitize.Media       = &Device->Media;
+  Device->MediaSanitize.MediaClear  = NvmExpressMediaClear;
+  Device->MediaSanitize.MediaPurge  = NvmExpressMediaPurge;
+  Device->MediaSanitize.MediaFormat = NvmExpressMediaFormat;
+
+  Device->Controller->Passthru.Mode             = 0;
+  Device->Controller->Passthru.PassThru         = NvmeDeviceUnitTestPassthru;
+  Device->Controller->Passthru.BuildDevicePath  = NULL;
+  Device->Controller->Passthru.GetNamespace     = NULL;
+  Device->Controller->Passthru.GetNextNamespace = NULL;
+
+  // ASSERT (
+  //  sizeof (Device->MediaSanitize.SanitizeCapabilities) ==
+  //  sizeof (Device->Controller->ControllerData->Sanicap)
+  //  );
+
+  // CopyMem (
+  //  &(Device->MediaSanitize.SanitizeCapabilities),
+  //  &(Device->Controller->ControllerData->Sanicap),
+  //  sizeof (Device->MediaSanitize.SanitizeCapabilities)
+  //  );
+
+  CopyMem (&Device->NamespaceData, NamespaceData, sizeof (NVME_ADMIN_NAMESPACE_DATA));
+  *ppDevice = Device;
+
+  DEBUG ((DEBUG_VERBOSE, "[          ] NvmeCreateDeviceInstance: Allocated and Initialized NVME_DEVICE_PRIVATE_DATA\n"));
+  DEBUG ((DEBUG_VERBOSE, "[          ] NvmeCreateDeviceInstance: Exit\n"));
+
+  return UNIT_TEST_PASSED;
+}
+
+UNIT_TEST_STATUS
+EFIAPI
+MediaSanitizePurgeUnitTest (
+  IN UNIT_TEST_CONTEXT  Context
+  )
+{
+  UINT32                    PurgeAction;
+  UINT32                    OverwritePattern;
+  UNIT_TEST_STATUS          UnitTestStatus = UNIT_TEST_PASSED;
+  NVME_DEVICE_PRIVATE_DATA  *NvmeDevice    = NULL;
+  EFI_STATUS                Status         = EFI_SUCCESS;
+
+  DEBUG ((DEBUG_VERBOSE, "[          ] MediaSanitizePurgeUnitTest: Starting Unit Test\n"));
+
+  UnitTestStatus = NvmeCreateDeviceInstance (&NvmeDevice);

```

Comment:
It could definitely be leveraged through the prerequisite and cleanup functions, but I was having issues with the memory allocation of the emulated NVMe device handling. Are you okay if I resolve this in another PR in the near future? I do agree this should be moved, I would just like to get the functional code part of this PR pushed first. @kuqin12

Reply:
Yes, agreed. Let's improve the unit test separately.